pkg/scheduler/podtask/port_mapping.go

```diff
@@ -2,168 +2,148 @@ package podtask
+	return portmap(offered, pod, cs, func(p *api.ContainerPort) bool {
+		return p.HostPort != 0
+	})
 }
 
-type DuplicateHostPortError struct {
-	m1, m2 HostPortMapping
+// WildcardPorts is the same as FixedPorts, except that .HostPorts of 0
+// are mapped to any offered port.
+func WildcardPorts(offered Ranges, pod string, cs ...api.Container) ([]PortMapping, error) {
+	return portmap(offered, pod, cs, func(_ *api.ContainerPort) bool {
+		return true
+	})
 }
 
-func (err *DuplicateHostPortError) Error() string {
-	return fmt.Sprintf(
-		"Host port %d is specified for container %d, pod %d and container %d, pod %d",
-		err.m1.OfferPort, err.m1.ContainerIdx, err.m1.PortIdx, err.m2.ContainerIdx, err.m2.PortIdx)
-}
+// byHostPort is an auxiliary type used to sort []PortMappings by their
+// HostPorts.
+type byHostPort []PortMapping
 
-// wildcard k8s host port mapping implementation: hostPort == 0 gets mapped to any available offer port
-func wildcardHostPortMapping(t *T, offer *mesos.Offer) ([]HostPortMapping, error) {
-	mapping, err := defaultHostPortMapping(t, offer)
-	if err != nil {
-		return nil, err
-	}
-	taken := make(map[uint64]struct{})
-	for _, entry := range mapping {
-		taken[entry.OfferPort] = struct{}{}
+func (pm byHostPort) Len() int           { return len(pm) }
+func (pm byHostPort) Less(i, j int) bool { return pm[i].HostPort < pm[j].HostPort }
+func (pm byHostPort) Swap(i, j int)      { pm[i], pm[j] = pm[j], pm[i] }
+
+// portmap maps matched task's container's host ports to offered ports.
+func portmap(offered Ranges, pod string, cs []api.Container, match func(*api.ContainerPort) bool) ([]PortMapping, error) {
+	if len(cs) == 0 {
+		return []PortMapping{}, nil
 	}
-	wildports := []HostPortMapping{}
-	for i, container := range t.Pod.Spec.Containers {
-		for pi, port := range container.Ports {
-			if port.HostPort == 0 {
-				wildports = append(wildports, HostPortMapping{
-					ContainerIdx: i,
-					PortIdx:      pi,
+
+	var wanted []PortMapping
+	for i := range cs {
+		for j, p := range cs[i].Ports {
+			if match(&p) {
+				wanted = append(wanted, PortMapping{
+					PodName:        pod,
+					ContainerIndex: i,
+					PortIndex:      j,
+					HostPort:       uint64(p.HostPort),
 				})
 			}
 		}
 	}
-	remaining := len(wildports)
-	foreachRange(offer, "ports", func(bp, ep uint64) {
-		log.V(3).Infof("Searching for wildcard port in range {%d:%d}", bp, ep)
-		for _, entry := range wildports {
-			if entry.OfferPort != 0 {
-				continue
-			}
-			for port := bp; port <= ep && remaining > 0; port++ {
-				if _, inuse := taken[port]; inuse {
-					continue
-				}
-				entry.OfferPort = port
-				mapping = append(mapping, entry)
-				remaining--
-				taken[port] = struct{}{}
-				break
-			}
+	// higher ports have highest priority, wildcard has the lowest
+	sort.Sort(sort.Reverse(byHostPort(wanted)))
+
+	taken := make(map[uint64]int, len(wanted))
+	for i := range wanted {
+		if offered.Size() < uint64(len(wanted)-len(taken)) {
```

Comment:
instead of referencing wanted[i].HostPort everywhere, wouldn't it be more readable to say:
```go
hostPort := &wanted[i].HostPort
```
and then reference that?